ocaml/xenopsd/c_stubs/xenctrlext_stubs.c

```diff
@@ -452,6 +452,96 @@ CAMLprim value stub_xenctrlext_cputopoinfo(value xch_val)
 	CAMLreturn(result);
 }
 
+/*
+ * Convert an Ocaml int64 array to a C uint32_t array, zero extending as
+ * necessary.
+ */
+static void ocaml_int64_array_to_c_array(value o, uint32_t *c, mlsize_t c_len)
+{
+	mlsize_t i, o_len = caml_array_length(o);
+
+	for (i = 0; i < o_len; ++i)
+		c[i] = Int64_val(Field(o, i));
+	for (; i < c_len; ++i)
+		c[i] = 0;
```

Comment:
This won't allow you to differentiate between: this entire feature leaf was not present before and got zero-extended, vs it was present before, but with a value of `0`. Especially with some features having opposite polarity "not present before" could actually mean "it was acting as 1 before", and zero-extending here would be wrong. Or do you intend to switch the polarity internally for any new bits that behave that way?

For backwards compatibility calculations the difference may matter at some point (and OCaml could make this distinction by looking at whether the array is empty or not).

Also do you have to special case the entirely zero-length featureset in case some OCaml code does a 'fold' with a starting value of an empty array? (otherwise that would result in the entire pool getting all zeros everywhere, and I think this was special cased before in the OCaml code), or is that already special cased in the caller?

Reply:
Discussed that the special fold case doesn't exist anymore and there is already a CPUID bit that can be used to solve the potential tri-state problem with the additional features (0,1, #GP).

Reply:
The zero extending is intentional, to avoid needing memory allocations in the library function, and therefore to avoid ENOMEM failures.

The new logic is not a case of having opposite polarity (I was hoping that was to be the case, but alas).

I could in principle pass `p1_orig_len` and `p2_orig_len` in which would allow distinguishing these cases, but I don't have any use them right now and I'm not sure it would be useful in the future either.

Current VMs don't see the ARCH_CAPS CPUID bit (and MSR_ARCH_CAPS will #GP on access).  There's no room for a VM to see an intermediate state with ARCH_CAPS visible but RSBA/RRSBA incorrectly calculated.